### PR TITLE
blob/all: add a Copy method BREAKING_CHANGE_OK

### PR DESCRIPTION
**Pull request by vangent**
cc @Zyqsempai 

This PR adds the portable API method, extends the driver interface, adds conformance tests, and implements `memblob` and `fileblob`.

It is marked as a breaking change due to the new method in the blob driver interface; since all drivers are in this repo this will not be a breaking change for our users.

Left for subsequent PRs (and marked with TODOs):

* Implementation for the 3 cloud provider drivers.
* Extending conformance tests for `As` to include `CopyOptions.BeforeCopy`.
* Verifying and documenting and testing the semantics for what happens if the destination already exists. Both `fileblob` and `memblob` overwrite, but I'm not sure what the cloud providers do yet.

Updates #1696.

### CODE REVIEW

File: blob/drivertest/drivertest.go [hunk 1]
```diff
@@ -1621,6 +1626,105 @@ func testMD5(t *testing.T, newHarness HarnessMaker) {
...
+		}
+		if !cmp.Equal(got, contents) {
+			t.Errorf("got %q want %q", string(got), string(contents))
+		}
+		// Verify attributes of the copy.
+		gotAttr, err := b.Attributes(ctx, dstKey)
+		if err != nil {
+			t.Error(err)
+		}
+		if diff := cmp.Diff(gotAttr, wantAttr, cmpopts.IgnoreUnexported(blob.Attributes{})); diff != "" {
```
Comment (jba) [inline blob/drivertest/drivertest.go]: Only compare if err == nil.

File: blob/fileblob/fileblob.go [hunk 2]
```diff
@@ -516,6 +516,48 @@ func (w *writer) Close() error {
...
+	srcPath, _, xa, err := b.forKey(srcKey)
+	if err != nil {
+		return err
+	}
+	f, err := os.Open(srcPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	r := io.Reader(f)
```
Comment (jba) [inline blob/fileblob/fileblob.go]: I don't think this is necessary.

File: blob/blob.go [hunk 3]
```diff
@@ -794,6 +795,37 @@ func (b *Bucket) NewWriter(ctx context.Context, key string, opts *WriterOptions)
...
+		return fmt.Errorf("blob.Copy: dstKey must be a valid UTF-8 string: %q", dstKey)
+	}
+	if opts == nil {
+		opts = &CopyOptions{}
+	}
+	dopts := &driver.CopyOptions{
+		BeforeCopy: opts.BeforeCopy,
+	}
+	b.mu.RLock()
+	defer b.mu.RUnlock()
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/blob.go]: This is following the pattern used in all the other APIs. I was worried about that too, but I think it's OK given that it's a read lock, and the only time we get the full/write lock is during `Close`.

**Comment by vangent** [comment]
> As a general note, I'd expect cross-bucket copying to be somewhat common. Is there a plan for that?

Good point. I haven't looked into that. Added a TODO.

File: blob/driver/driver.go [hunk 4]
```diff
@@ -255,6 +263,16 @@ type Bucket interface {
 	// and do any necessary cleanup in Close. Close should then return ctx.Err().
 	NewTypedWriter(ctx context.Context, key string, contentType string, opts *WriterOptions) (Writer, error)
```
[the remark this answers is not in the document]
Comment (vangent) [reply blob/driver/driver.go]: Done.

**Comment by vangent** [comment]
> > As a general note, I'd expect cross-bucket copying to be somewhat common. Is there a plan for that?

My tentative plan would be to add `CopyOptions.SourceBucket *blob.Bucket`. The default (`nil`) means use the same bucket, but you could supply another bucket to copy from.

File: blob/fileblob/fileblob.go (hunk 2 above)
Comment (vangent) [reply blob/fileblob/fileblob.go]: Done.

File: blob/drivertest/drivertest.go (hunk 1 above)
Comment (vangent) [reply blob/drivertest/drivertest.go]: Changed `t.Error` to `t.Fatal` above.

File: blob/fileblob/fileblob.go [hunk 5]
```diff
@@ -516,6 +516,47 @@ func (w *writer) Close() error {
 	return nil
 }
 
+// Copy implements driver.Copy.
+func (b *bucket) Copy(ctx context.Context, srcKey, dstKey string, opts *driver.CopyOptions) error {
+	// Note: we could use NewRangedReader here, but since we need to copy all of
+	// the metadata (from xa), it's more efficient to do it directly.
+	srcPath, _, xa, err := b.forKey(srcKey)
```
Comment (clausti) [inline blob/fileblob/fileblob.go]: is `xa` a conventional abbreviation/var name for metadata?

Comment (vangent) [reply blob/fileblob/fileblob.go]: I believe it's a shortening of the type (`type xattrs struct`), which I think is short for "extended attributes". `xa` is used in a number of other places in this file though. I'm open to better names, but probably in a different PR.

**Comment by vangent** [comment]
> > > As a general note, I'd expect cross-bucket copying to be somewhat common. Is there a plan for that?
> 
> My tentative plan would be to add `CopyOptions.SourceBucket *blob.Bucket`. The default (`nil`) means use the same bucket, but you could supply another bucket to copy from.

Filed https://github.com/google/go-cloud/issues/1766.

File: blob/blob.go [hunk 6]
```diff
@@ -792,6 +793,36 @@ func (b *Bucket) NewWriter(ctx context.Context, key string, opts *WriterOptions)
...
 }
 
+// Copy the blob stored at srcKey to dstKey.
+// A nil CopyOptions is treated the same as the zero value.
+//
+// If the source blob does not exist, Copy returns an error for which
+// gcerrors.Code will return gcerrors.NotFound.
+//
+// If the destination blob already exists, it is overwritten.
+func (b *Bucket) Copy(ctx context.Context, srcKey, dstKey string, opts *CopyOptions) (err error) {
```
Comment (jba) [inline blob/blob.go]: I think you should reverse src and dest, to be compatible with Go (e.g. the `copy` built-in).

Comment (vangent) [reply blob/blob.go]: Good point, done.

**Comment by vangent** [comment]
PTAL

**Pull request merged**